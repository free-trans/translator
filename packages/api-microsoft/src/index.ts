import type { LanguageCode, TranslateQueryResult } from '@arvinxu/translator';
import { RequestErrorCode, Translator } from '@arvinxu/translator';

import { langMap } from './langMap';

type MicrosoftRawResult = MicrosoftErrorResult | MicrosoftSuccessResult;

interface MicrosoftErrorResult {
  error: {
    code: 401000 | number;
    message: string;
  };
}
type MicrosoftSuccessResult = [
  {
    detectedLanguage?: {
      language: string;
      score: number;
    };
    translations: [{ text: string; to: string }];
  },
];

/**
 * 微软翻译引擎配置
 */
export interface MicrosoftConfig {
  token?: string;
  /**
   * 秘钥
   * @see https://docs.microsoft.com/zh-cn/azure/cognitive-services/translator/reference/v3-0-reference
   */
  secretKey?: string;
  /**
   * 翻译器资源是翻译器资源的区域
   * @description 如果使用区域翻译器资源调用 Translator,
   * 那么就需要配置这个参数
   *
   * @example
   * ```bash
   * // Pass secret key and region using headers
   * curl -X POST "https://api.cognitive.microsofttranslator.com/translate?api-version=3.0&to=es" \
   * -H "Ocp-Apim-Subscription-Key:<your-secret-key>" \
   * -H "Ocp-Apim-Subscription-Region:<your-region>" \
   * -H "Content-Type: application/json" \
   * -d "[{'Text':'Hello, what is your name?'}]"
   * ```
   * @see https://docs.microsoft.com/zh-cn/azure/cognitive-services/translator/reference/v3-0-reference
   */
  region?: string;
}

export class Microsoft extends Translator<MicrosoftConfig> {
  /** Translator lang to custom lang */
  private static readonly langMap = new Map(langMap);

  /** Custom lang to translator lang */
  private static readonly langMapReverse = new Map(
    langMap.map(([translatorLang, lang]) => [lang, translatorLang]),
  );

  userAgent =
    'Mozilla/5.0 (Linux; Android 5.0; SM-G900P Build/LRX21T) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/69.0.3497.100 Mobile Safari/537.36';

  private token: string;
  /**
   * 请求
   * @param text
   * @param from
   * @param to
   * @param token
   * @private
   */
  private async fetchWithToken(
    /** 'com' or 'cn' */
    text: string,
    from: LanguageCode,
    to: LanguageCode,
    token: string,
  ): Promise<TranslateQueryResult> {
    this.token = token;

    const url = 'https://api.cognitive.microsofttranslator.com/translate';

    const fetch = async () => {
      return this.request.post<MicrosoftRawResult>(url, {
        params: {
          from: from === 'auto' ? null : from,
          to,
          'api-version': '3.0',
          includeSentenceLength: false,
        },
        data: [{ Text: text }],
        headers: {
          Authorization: `Bearer ${this.token}`,
          'User-Agent': this.userAgent,
        },
      });
    };

    const data = await fetch()
      // 如果第一次请求有错误
      .catch(async () => {
        // 更新一次 token
        await this.updateToken();

        return await fetch().catch((e) => {
          // 第二次请求有错误的话返回错误结果
          return e.data as MicrosoftErrorResult;
        });
      });

    // 处理错误状态
    if (!(data instanceof Array)) {
      const { message, code } = data.error;
      switch (code) {
        case 401000:
          return this.getErrorResult(RequestErrorCode.UNAUTHORIZED, message);
        default:
          return this.getErrorResult(code, message);
      }
    }
    const { detectedLanguage, translations } = data[0];

    const { text: trans, to: _to } = translations[0];

    return {
      success: true,
      origin: {
        paragraphs: text.split('\n'),
      },
      trans: {
        paragraphs: trans.split('\n'),
      },
      text,
      from: Microsoft.langMapReverse.get(detectedLanguage?.language) || from,
      to: Microsoft.langMapReverse.get(_to),
    };
  }

  async updateToken(): Promise<void> {
    const token = await this.request.get<string>(
      'https://edge.microsoft.com/translate/auth',
      {
        headers: {
          'User-Agent': this.userAgent,
        },
      },
    );

    if (typeof token === 'string') {
      this.token = token;
    }
  }

  protected async query(
    text: string,
    from: LanguageCode,
    to: LanguageCode,
    config: MicrosoftConfig,
  ): Promise<TranslateQueryResult> {
    let { token } = this.config;
    if (config.token) {
      token = config.token;
    }
    const data = await this.fetchWithToken(text, from, to, token);

    return data;
  }

  readonly name = 'microsoft';

  getSupportLanguages(): LanguageCode[] {
    return [...Microsoft.langMap.keys()];
  }
}

export default Microsoft;
