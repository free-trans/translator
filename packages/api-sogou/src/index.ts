import {
  LanguageCode,
  Translator,
  TranslateQueryResult,
  RequestErrorCode,
  TranslateResult,
} from '@arvinxu/translator';
import md5 from 'md5';

import { langMap } from './langMap';

type SogouErrorCode = string

type SogouTranslateError = {
  error_code: SogouErrorCode;
  error_msg: 'Invalid Sign' | string;
};

type SogouTranslateResult = {
  from: string;
  to: string;
  translate: {
    dst: string;
    orig_text: string;
  }[];
};

export interface SogouConfig {
  uuid: string;
}

export class Sogou extends Translator<SogouConfig> {
  readonly name = 'sogou';

  readonly endpoint = 'https://fanyi.sogou.com/api/transpc/text/result';

  /**
   * 获取结果方法
   * @param text
   * @param from
   * @param to
   * @param config
   * @protected
   */
  protected async query(
    text: string,
    from: LanguageCode,
    to: LanguageCode,
    config: SogouConfig,
  ): Promise<TranslateQueryResult> {
    const { uuid } = config;

    const data = await this.request.get<SogouTranslateResult | SogouTranslateError>(
      this.endpoint, {
        responseType: 'json',
        params: {
          from: Sogou.langMap.get(from),
          to: Sogou.langMap.get(to),
          q: text,
          uuid,
          //需优化
          sign: md5(from + to + text),
        },
      });

    // 如果报错 则返回错误信息
    if ((data as SogouTranslateError).error_code) {
      const { error_code, error_msg } = data as SogouTranslateError;

      return this.getErrorResult(this.handleErrorCode(error_code), error_msg);
    }

    const {
      translate: transResult,
      from: langDetected,
    } = data as SogouTranslateResult;

    const transParagraphs = transResult.map(({ dst }) => dst);
    const detectedFrom = Sogou.langMapReverse.get(langDetected) as LanguageCode;

    return {
      success: true,
      text,
      from: detectedFrom,
      to,
      origin: {
        paragraphs: transResult.map(({ orig_text }) => orig_text),
      },
      trans: {
        paragraphs: transParagraphs,
      },
    };
  }

  /**
   * 解析错误码
   * @param code
   */
  private handleErrorCode(code: SogouErrorCode) {
    switch (code) {
      case '52003':
        return RequestErrorCode.UNAUTHORIZED;
      case '54003':
        return RequestErrorCode.LIMITED;
      default:
        return parseInt(code, 10);
    }
  }

  /** Translator lang to custom lang */
  private static readonly langMap = new Map(langMap);

  /** Custom lang to translator lang */
  private static readonly langMapReverse = new Map(
    langMap.map(([translatorLang, lang]) => [lang, translatorLang]),
  );


  getSupportLanguages(): LanguageCode[] {
    return [...Sogou.langMap.keys()];
  }

  // async detect(text: string): Promise<Language> {
  // }

  // async textToSpeech(text: string, lang: LanguageCode): Promise<string | null> {
  // }
}

export default Sogou;
