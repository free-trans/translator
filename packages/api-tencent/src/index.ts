import type { LanguageCode, TranslateQueryResult } from '@arvinxu/translator';
import { Translator } from '@arvinxu/translator';

import { decodeHTMLEntities } from './utils';
import { langMap } from './langMap';
import { sdkTranslate } from './sdk';
import type { TencentConfig, TencentResponseResult } from './type';

export { TencentConfig } from './type';

export class Tencent extends Translator<TencentConfig> {
  /** Translator lang to custom lang */
  private static readonly langMap = new Map(langMap);

  /** Custom lang to translator lang */
  private static readonly langMapReverse = new Map(
    langMap.map(([translatorLang, lang]) => [lang, translatorLang]),
  );

  /**
   * tencent 获取 token 的地址
   * @private
   */
  private tokenUrl: string;

  private qtk: string;
  private qtv: string;

  private updateAt: number;

  /**
   * 使用 SDK 请求
   * @param text
   * @param from
   * @param to
   * @param config
   * @private
   */
  private async requestWithSDK(
    text: string,
    from: LanguageCode,
    to: LanguageCode,
    config: TencentConfig,
  ) {
    const data = await sdkTranslate(
      text,
      Tencent.langMap.get(from),
      Tencent.langMap.get(to),
      {
        secretId: config.secretId,
        secretKey: config.secretKey,
        region: config.region,
      },
    );

    return {
      success: true,
      text,
      from: Tencent.langMapReverse.get(data.Source) || from,
      to: Tencent.langMapReverse.get(data.Target) || to,
      origin: {
        paragraphs: text.split(/\n+/),
      },
      trans: {
        paragraphs: decodeHTMLEntities(data.TargetText).split(/\n+/),
      },
    };
  }

  protected async query(
    text: string,
    from: LanguageCode,
    to: LanguageCode,
    config?: TencentConfig,
  ): Promise<TranslateQueryResult> {
    if (config && config.withSDK) {
      return await this.requestWithSDK(text, from, to, config);
    }

    return await this.requestWithoutSDK(text, from, to);
  }

  /**
   * 不使用 SDK 请求
   * @param text
   * @param from
   * @param to
   * @private
   */
  private async requestWithoutSDK(
    text: string,
    from: LanguageCode,
    to: LanguageCode,
  ): Promise<TranslateQueryResult> {
    await this.fetchToken();

    const { translate } = await this.request.post<TencentResponseResult>(
      'https://fanyi.qq.com/api/translate',
      {
        requestType: 'form',
        data: {
          source: Tencent.langMap.get(from),
          target: Tencent.langMap.get(to),
          sourceText: text,
          qtv: this.qtv,
          qtk: this.qtk,
        },
        headers: {
          'User-Agent':
            'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.150 Safari/537.36',
          Origin: 'https://fanyi.qq.com',
          Connection: 'keep-alive',
        },
      },
    );

    const formatResult = (raw: string[]) => {
      // 腾讯翻译器的结果结构如下
      //   [
      //   {
      //     sourceText: '你好',
      //     targetText: 'Hello. ',
      //   },
      //     {
      //       sourceText: '\n',
      //       targetText: '\n ',
      //     },
      //     {
      //       sourceText: '我爱你',
      //       targetText: 'I love you',
      //     },
      //   ];
      // 因此采用合并然后按 \n 切分的思路处理段落问题

      return raw
        .join('')
        .split('\n')
        .map((i) => i.trim());
    };

    return {
      success: true,
      from: Tencent.langMapReverse.get(translate.source) || from,
      to: Tencent.langMapReverse.get(translate.target) || to,
      text,
      trans: {
        paragraphs: formatResult(translate.records.map((r) => r.targetText)),
      },
      origin: {
        paragraphs: formatResult(translate.records.map((r) => r.sourceText)),
      },
    };
  }

  async fetchToken() {
    if (!this.tokenUrl) {
      await this.fetchTokenUrl();
    }

    const data = await this.request
      .post<{ qtk: string; qtv: string }>(
        `https://fanyi.qq.com/api/${this.tokenUrl}`,
      )
      .catch((e) => {
        console.error(e);
        return { qtv: '', qtk: '' };
      });

    const { qtk, qtv } = data;
    this.qtk = qtk;
    this.qtv = qtv;
    this.updateAt = Date.now();

    return { qtk, qtv };
  }
  /**
   * 获取请求 token 的 url
   */
  async fetchTokenUrl() {
    const data = await this.request('https://fanyi.qq.com/');

    const reauthuri = /reauthuri\s=\s"(.*)"/.exec(data);

    // eslint-disable-next-line prefer-destructuring
    this.tokenUrl = reauthuri[1];

    return reauthuri[1];
  }

  readonly name = 'tencent';

  getSupportLanguages(): LanguageCode[] {
    return [...Tencent.langMap.keys()];
  }
}

export default Tencent;
