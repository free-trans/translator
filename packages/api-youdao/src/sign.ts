import { sha256 } from 'js-sha256';
import md5 from 'md5';

/**
 * 有道接口的签名算法
 * @param text 待签名文本
 * @param token
 * @param userAgent
 */
export const signKey = (text: string, token: string, userAgent: string) => {
  const client = 'fanyideskweb';

  if (!userAgent)
    userAgent =
      '5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.149 Safari/537.36';

  const lts = `${Date.now()}`;

  const salt = lts + parseInt(`${Math.random() * 10}`, 10);

  const bv = md5(userAgent);
  return {
    client,
    lts,
    bv,
    salt,
    sign: md5(client + text + salt + token),
  };
};

/**
 * 对签名后的数据做一些处理
 * @param q
 */
const truncate = (q: string): string => {
  const len = q.length;
  if (len <= 20) return q;
  return q.substring(0, 10) + len + q.substring(len - 10, len);
};

export const signAPI = (text: string, appKey: string, key: string) => {
  const salt = new Date().getTime();
  const curtime = Math.round(new Date().getTime() / 1000);
  const str1 = appKey + truncate(text) + salt + curtime + key;
  const sign = sha256(str1);

  return {
    appKey,
    salt,
    curtime,
    sign,
  };
};
