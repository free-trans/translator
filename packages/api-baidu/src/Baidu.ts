import type {
  LanguageCode,
  TranslateQueryResult,
  TranslateResult,
} from '@arvinxu/translator';
import { RequestErrorCode, Translator } from '@arvinxu/translator';
import md5 from 'md5';

import { langMap } from './langMap';
import sign from './utils/sign';
import {
  BaiduConfig,
  BaiduErrorCode,
  BaiduFreeTranslateError,
  BaiduFreeTranslateResult,
  BaiduTranslateError,
  BaiduTranslateResult,
} from './types';

export class Baidu extends Translator<BaiduConfig> {
  readonly name = 'baidu';

  readonly endpoint = 'https://fanyi.baidu.com/v2transapi';
  /**
   * 收费版本接口
   */
  readonly vipEndpoint = 'https://api.fanyi.baidu.com/api/trans/vip/translate';

  /**
   * 获取结果方法
   * @param text
   * @param from
   * @param to
   * @param config
   * @protected
   */
  protected async query(
    text: string,
    from: LanguageCode,
    to: LanguageCode,
    config: BaiduConfig,
  ): Promise<TranslateQueryResult> {
    if (config.free) return this.freeRequest(text, from, to, config.free);

    return this.apiRequest(text, from, to, config);
  }

  /**
   * 调用的接口
   */
  async translate(
    text: string,
    from: LanguageCode,
    to: LanguageCode,
    config?: BaiduConfig,
  ): Promise<TranslateResult> {
    let fetchCount = 0;

    // 提取请求方法
    const fetch = async () => {
      return this.query(text, from, to, {
        ...this.config,
        ...config,
      });
    };

    let data = await fetch();

    while (data.success === false && data.code === RequestErrorCode.LIMITED) {
      if (fetchCount <= 2) {
        fetchCount += 1;
        this.sleep(100);
        // eslint-disable-next-line no-await-in-loop
        data = await fetch();
      }
    }

    return {
      type: this.name,
      text,
      from,
      to,
      ...data,
    };
  }

  /**
   * 解析错误码
   * @param code
   */
  private handleErrorCode(code: BaiduErrorCode) {
    switch (code) {
      case '52003':
        return RequestErrorCode.UNAUTHORIZED;
      case '54003':
        return RequestErrorCode.LIMITED;
      default:
        return parseInt(code, 10);
    }
  }
  /** Translator lang to custom lang */
  private static readonly langMap = new Map(langMap);

  /** Custom lang to translator lang */
  private static readonly langMapReverse = new Map(
    langMap.map(([translatorLang, lang]) => [lang, translatorLang]),
  );

  getSupportLanguages(): LanguageCode[] {
    return [...Baidu.langMap.keys()];
  }

  private sleep = (milliseconds) => {
    const date = Date.now();
    let currentDate = null;
    do {
      currentDate = Date.now();
    } while (currentDate - date < milliseconds);
  };

  /**
   *
   * @param text
   * @param from
   * @param to
   * @param config
   */
  private apiRequest = async (
    text: string,
    from: LanguageCode,
    to: LanguageCode,
    config: BaiduConfig,
  ) => {
    const salt = Date.now();
    const { appid, key } = config;
    const data = this.request.get<BaiduTranslateResult | BaiduTranslateError>(
      this.vipEndpoint,
      {
        requestType: 'json',
        params: {
          from: Baidu.langMap.get(from),
          to: Baidu.langMap.get(to),
          q: text,
          salt,
          appid,
          sign: md5(appid + text + salt + key),
        },
      },
    );

    if ((data as unknown as BaiduTranslateError).error_code) {
      const { error_code, error_msg } = data as unknown as BaiduTranslateError;

      return this.getErrorResult(this.handleErrorCode(error_code), error_msg);
    }

    const { trans_result: transResult, from: langDetected } =
      data as unknown as BaiduTranslateResult;

    const transParagraphs = transResult.map(({ dst }) => dst);
    const detectedFrom = Baidu.langMapReverse.get(langDetected) as LanguageCode;

    return {
      success: true,
      text,
      from: detectedFrom,
      to,
      origin: {
        paragraphs: transResult.map(({ src }) => src),
      },
      trans: {
        paragraphs: transParagraphs,
      },
    };
  };

  private freeRequest = async (
    text: string,
    from: LanguageCode,
    to: LanguageCode,
    config: BaiduConfig['free'],
  ) => {
    const res = await this.request.get<
      BaiduFreeTranslateResult | BaiduFreeTranslateError
    >(this.endpoint, {
      requestType: 'form',
      params: {
        from: Baidu.langMap.get(from),
        to: Baidu.langMap.get(to),
        query: text,
        transtype: 'translang',
        simple_means_flag: 3,
        domain: 'common',
        //
        token: config.token,
        sign: sign(text, config.seed),
      },
      headers: {
        'X-Requested-With': 'XMLHttpRequest',
        Cookie: config.cookie,
      },
    });

    // 如果报错 则返回错误信息
    if ((res as BaiduFreeTranslateError).errno) {
      const { errno, errmsg } = res as BaiduFreeTranslateError;

      return this.getErrorResult(this.handleErrorCode(errno), errmsg);
    }

    const { data: transResult, from: langDetected } = (
      res as BaiduFreeTranslateResult
    ).trans_result;

    const transParagraphs = transResult.map(({ dst }) => dst);
    const detectedFrom = Baidu.langMapReverse.get(langDetected) as LanguageCode;

    return {
      success: true,
      text,
      from: detectedFrom,
      to,
      origin: {
        paragraphs: transResult.map(({ src }) => src),
      },
      trans: {
        paragraphs: transParagraphs,
      },
    };
  };
}
