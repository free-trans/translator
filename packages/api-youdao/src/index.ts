import type { LanguageCode, TranslateQueryResult } from '@arvinxu/translator';
import { RequestErrorCode, Translator } from '@arvinxu/translator';
import { langMap } from './langMap';
import { signAPI, signKey } from './sign';

export interface YoudaoConfig {
  appKey?: string;
  key?: string;
  onlyAPI?: boolean;
  token?: string;
  userAgent?: string;
  cookie?: string;
}

interface YoudaoTranslateResult {
  errorCode: string;
  query: string;
  translation: string[];
  l: string;
  requestId?: string;
}

export interface YoudaoResponse {
  errorCode: number;
  translateResult: {
    src: string;
    tgt: string;
  }[][];
  /**
   * 翻译语言
   */
  type: string;
}

export class Youdao extends Translator<YoudaoConfig> {
  /** Translator lang to custom lang */
  private static readonly langMap = new Map(langMap);

  /** Custom lang to translator lang */
  private static readonly langMapReverse = new Map(
    langMap.map(([translatorLang, lang]) => [lang, translatorLang]),
  );

  /**
   * 不用 token 的请求方法
   * @param text
   * @param from
   * @param to
   * @param config
   * @private
   */
  private async fetchWithToken(
    text: string,
    from: LanguageCode,
    to: LanguageCode,
    config: YoudaoConfig,
  ) {
    const baseURL = 'http://fanyi.youdao.com';
    const url = `${baseURL}/translate_o?smartresult=dict&smartresult=rule`;

    if (!config.token) {
      return this.getErrorResult(
        RequestErrorCode.UNAUTHORIZED,
        'UNAUTHORIZED USER',
      );
    }
    const sign = signKey(text, config.token, config.userAgent); // 签名

    const data = {
      i: text,
      from,
      to,
      smartresult: 'dict',
      doctype: 'json',
      version: '2.1',
      keyfrom: 'fanyi.web',
      action: 'FY_BY_CLICKBUTTION',
      typoResult: 'false',
      ...sign,
    };

    // 发请求
    const res = await this.request.post<YoudaoResponse>(url, {
      data,
      requestType: 'form',
      headers: {
        Referer: baseURL,
        Cookie:
          config.cookie || 'OUTFOX_SEARCH_USER_ID=-1799202939@10.168.8.64',
      },
    });

    // 请求后处理
    if (!res.translateResult) {
      const { errorCode } = res;

      switch (errorCode) {
        case 50:
          return this.getErrorResult(
            RequestErrorCode.BAD_REQUEST,
            '入参信息有误',
          );
        default:
          return this.getErrorResult(errorCode, '错误');
      }
    }

    const { type, translateResult } = res;
    const [_from, _to] = type.split('2');

    const origin = translateResult.map((p) => p.map((i) => i.src).join(''));
    const trans = translateResult.map((p) => p.map((i) => i.tgt).join(''));

    return {
      success: true,
      text,
      from: Youdao.langMapReverse.get(_from),
      to: Youdao.langMapReverse.get(_to),
      origin: {
        paragraphs: origin,
      },
      trans: {
        paragraphs: trans,
      },
    };
  }

  /**
   * 利用 token 的请求方法
   * @param text
   * @param from
   * @param to
   * @param config
   * @private
   */
  private async fetchWithoutToken(
    text: string,
    from: LanguageCode,
    to: LanguageCode,
    config?: YoudaoConfig,
  ): Promise<TranslateQueryResult> {
    const { key, appKey } = config;

    const endpoint = 'http://openapi.youdao.com/api';
    if (!(key && appKey)) {
      return this.getErrorResult(401, 'NO TOKEN');
    }

    const signature = signAPI(text, config.appKey, config.key);

    // 生成参数
    const params = {
      q: text,
      from,
      to,
      signType: 'v3',
      ...signature,
    };

    const result = await this.request.get<YoudaoTranslateResult>(endpoint, {
      params,
    });

    // 错误处理
    switch (result.errorCode) {
      case '108':
        return this.getErrorResult(
          RequestErrorCode.UNAUTHORIZED,
          'UNAUTHORIZED USER',
        );
      case '0':
        break;
      default:
        return this.getErrorResult(
          parseInt(result.errorCode, 10),
          String(result),
        );
    }

    const [_from, _to] = result.l.split('2');

    return {
      success: true,
      text,
      from: Youdao.langMapReverse.get(_from),
      to: Youdao.langMapReverse.get(_to),
      origin: {
        paragraphs: text.split(/\n+/),
      },
      trans: {
        paragraphs: result.translation,
      },
    };
  }

  protected async query(
    text: string,
    from: LanguageCode,
    to: LanguageCode,
    config?: YoudaoConfig,
  ): Promise<TranslateQueryResult> {
    const finalConfig = {
      ...this.config,
      ...config,
    };

    const { appKey, key, onlyAPI } = finalConfig;

    // 只使用 API
    if (onlyAPI) {
      return await this.fetchWithoutToken(text, from, to, { appKey, key });
    }

    const data = await this.fetchWithToken(text, from, to, finalConfig);

    if (data.success) return data;

    // 如果有 appKey 和 key 则尝试发送 api 请求
    if (appKey && key) {
      return await this.fetchWithoutToken(text, from, to, {
        appKey,
        key,
      });
    }

    return data;
  }

  readonly name = 'youdao';

  getSupportLanguages(): LanguageCode[] {
    return [...Youdao.langMap.keys()];
  }
}

export default Youdao;
