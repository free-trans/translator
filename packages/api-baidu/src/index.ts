import { Baidu } from './Baidu';

export default Baidu;
export { Baidu } from './Baidu';
export * from './utils/crackUtils';
export { BaiduConfig } from './types';
