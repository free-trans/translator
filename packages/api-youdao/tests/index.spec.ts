import { Youdao } from '../src';

describe('有道翻译接口', () => {
  // API 地址 http://api.fanyi.baidu.com/api/trans/product/prodinfo

  const youdao = new Youdao({
    config: {
      cookie: process.env.YOUDAO_COOKIE,
      token: process.env.YOUDAO_TOKEN,
      userAgent: process.env.YOUDAO_USER_AGENT,
      appKey: process.env.YOUDAO_APP_ID,
      key: process.env.YOUDAO_KEY,
    },
  });

  it('should get supported languages', () => {
    const result = youdao.getSupportLanguages();

    expect(result).toContain('auto');
    expect(result).toContain('zh-CN');
    expect(result).toContain('en');
  }, 5000);

  describe('正常调用', () => {
    it('英翻中正常', async () => {
      const data = await youdao.translate('I love you', 'auto', 'zh-CN');

      expect(data).toEqual({
        success: true,
        type: 'youdao',
        text: 'I love you',
        from: 'en',
        to: 'zh-CN',
        /** 原文 */
        origin: {
          paragraphs: ['I love you'],
        },
        /** 译文 */
        trans: {
          paragraphs: ['我爱你'],
        },
      });
    }, 9000);

    it('中翻英正常', async () => {
      const data = await youdao.translate('我爱你', 'auto', 'en');

      expect(data).toEqual({
        success: true,
        type: 'youdao',
        text: '我爱你',
        from: 'zh-CN',
        to: 'en',
        /** 原文 */
        origin: {
          paragraphs: ['我爱你'],
        },
        /** 译文 */
        trans: {
          paragraphs: ['I love you'],
        },
      });
    }, 9000);

    it('中翻英正常', async () => {
      const data = await youdao.translate('我爱你', 'auto', 'en');

      expect(data).toEqual({
        success: true,
        type: 'youdao',
        text: '我爱你',
        from: 'zh-CN',
        to: 'en',
        /** 原文 */
        origin: {
          paragraphs: ['我爱你'],
        },
        /** 译文 */
        trans: {
          paragraphs: ['I love you'],
        },
      });
    }, 9000);

    it('token 不对时正常', async () => {
      const data = await youdao.translate('我爱你', 'auto', 'en', {
        token: '',
      });

      expect(data).toEqual({
        success: true,
        type: 'youdao',
        text: '我爱你',
        from: 'zh-CN',
        to: 'en',
        /** 原文 */
        origin: {
          paragraphs: ['我爱你'],
        },
        /** 译文 */
        trans: {
          paragraphs: ['I love you'],
        },
      });
    }, 9000);
  });

  describe('纯 API', () => {
    it('英翻中正常', async () => {
      const data = await youdao.translate('I love you', 'auto', 'zh-CN', {
        onlyAPI: true,
      });

      expect(data).toEqual({
        success: true,
        type: 'youdao',
        text: 'I love you',
        from: 'en',
        to: 'zh-CN',
        /** 原文 */
        origin: {
          paragraphs: ['I love you'],
        },
        /** 译文 */
        trans: {
          paragraphs: ['我爱你'],
        },
      });
    }, 9000);

    it('长句结果正常', async () => {
      const data = await youdao.translate(
        'Webpack is an open-source JavaScript module bundler. It is made primarily for JavaScript, but it can transform front-end assets such as HTML, CSS, and images if the corresponding loaders are included. webpack takes modules with dependencies and generates static assets representing those modules.',
        'auto',
        'zh-CN',
        {
          onlyAPI: true,
        },
      );
      expect(data).toEqual({
        type: 'youdao',
        success: true,
        text:
          'Webpack is an open-source JavaScript module bundler. It is made primarily for JavaScript, but it can transform front-end assets such as HTML, CSS, and images if the corresponding loaders are included. webpack takes modules with dependencies and generates static assets representing those modules.',
        to: 'zh-CN',
        from: 'en',
        origin: {
          paragraphs: [
            'Webpack is an open-source JavaScript module bundler. It is made primarily for JavaScript, but it can transform front-end assets such as HTML, CSS, and images if the corresponding loaders are included. webpack takes modules with dependencies and generates static assets representing those modules.',
          ],
        },
        trans: {
          paragraphs: [
            'Webpack是一个开源的JavaScript模块捆绑器。它主要是为JavaScript制作的，但它可以转换前端资产，如HTML、CSS和图像(如果包含相应的加载器的话)。webpack接受带有依赖项的模块，并生成代表这些模块的静态资产。',
          ],
        },
      });
    });

    it('未初始化时报 401 错', async () => {
      const data = await youdao.translate('I love you', 'auto', 'zh-CN', {
        cookie: '',
        token: '',
        userAgent: '',
        appKey: '',
        key: '',
        onlyAPI: true,
      });
      expect(data.success).toBeFalsy();
      expect(data.code).toBe(401);
    });
  });

  describe('withToken', () => {
    const config = {
      cookie: process.env.YOUDAO_COOKIE,
      token: process.env.YOUDAO_TOKEN,
      userAgent: process.env.YOUDAO_USER_AGENT,
    };

    const errorResult = {
      code: 400,
      message: '入参信息有误',
      origin: {
        paragraphs: [],
      },
      success: false,
      trans: {
        paragraphs: [],
      },
    };
    it('正常获得结果', async () => {
      const data = await youdao['fetchWithToken'](
        'I love you',
        'auto',
        'zh-CN',
        config,
      );
      expect(data).toEqual({
        success: true,
        text: 'I love you',
        to: 'zh-CN',
        from: 'en',
        origin: {
          paragraphs: ['I love you'],
        },
        trans: {
          paragraphs: ['我爱你'],
        },
      });
    });
    it('长句结果正常', async () => {
      const data = await youdao['fetchWithToken'](
        'Webpack is an open-source JavaScript module bundler. It is made primarily for JavaScript, but it can transform front-end assets such as HTML, CSS, and images if the corresponding loaders are included. webpack takes modules with dependencies and generates static assets representing those modules.',
        'auto',
        'zh-CN',
        config,
      );
      expect(data).toEqual({
        success: true,
        text:
          'Webpack is an open-source JavaScript module bundler. It is made primarily for JavaScript, but it can transform front-end assets such as HTML, CSS, and images if the corresponding loaders are included. webpack takes modules with dependencies and generates static assets representing those modules.',
        to: 'zh-CN',
        from: 'en',
        origin: {
          paragraphs: [
            'Webpack is an open-source JavaScript module bundler. It is made primarily for JavaScript, but it can transform front-end assets such as HTML, CSS, and images if the corresponding loaders are included. webpack takes modules with dependencies and generates static assets representing those modules.',
          ],
        },
        trans: {
          paragraphs: [
            'Webpack是一个开源的JavaScript模块捆绑器。它主要是为JavaScript制作的，但它可以转换前端资产，如HTML、CSS和图像(如果包含相应的加载器的话)。webpack接受带有依赖项的模块，并生成代表这些模块的静态资产。',
          ],
        },
      });
    });
    it('段落翻译正常', async () => {
      const data = await youdao['fetchWithToken'](
        'Webpack is an open-source JavaScript module bundler.\n It is made primarily for JavaScript, but it can transform front-end assets such as HTML, CSS, and images if the corresponding loaders are included. webpack takes modules with dependencies and generates static assets representing those modules.',
        'auto',
        'zh-CN',
        config,
      );
      expect(data.success).toBeTruthy();
      expect(data.origin.paragraphs).toHaveLength(2);
      expect(data.trans.paragraphs).toHaveLength(2);
    });

    describe('报错', () => {
      it('无内容调用报错', async () => {
        const data = await youdao['fetchWithToken']('', 'en', 'zh-CN', config);

        expect(data).toStrictEqual(errorResult);
      });
      it('入参不正确时报错', async () => {
        const data = await youdao['fetchWithToken'](
          'I love you',
          'auto',
          'zh-CN',
          {
            cookie: '',
            token: '',
            userAgent: '',
          },
        );
        expect(data).toEqual({
          ...errorResult,
          code: 401,
          message: 'UNAUTHORIZED USER',
        });
      });
    });
  });

  describe('初始化方法', () => {
    it('只入参 token 时正常', async () => {
      const youdao = new Youdao({
        config: {
          token: process.env.YOUDAO_TOKEN,
        },
      });

      const data = await youdao.translate('I love you', 'auto', 'zh-CN');

      expect(data).toEqual({
        success: true,
        type: 'youdao',
        text: 'I love you',
        from: 'en',
        to: 'zh-CN',
        /** 原文 */
        origin: {
          paragraphs: ['I love you'],
        },
        /** 译文 */
        trans: {
          paragraphs: ['我爱你'],
        },
      });
    }, 9000);
  });
});
