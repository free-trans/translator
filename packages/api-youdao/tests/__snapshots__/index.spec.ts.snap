// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`有道翻译接口 withToken 长句结果正常 1`] = `
Object {
  "from": "en",
  "origin": Object {
    "paragraphs": Array [
      "Webpack is an open-source JavaScript module bundler. It is made primarily for JavaScript, but it can transform front-end assets such as HTML, CSS, and images if the corresponding loaders are included. webpack takes modules with dependencies and generates static assets representing those modules.",
    ],
  },
  "success": true,
  "text": "Webpack is an open-source JavaScript module bundler. It is made primarily for JavaScript, but it can transform front-end assets such as HTML, CSS, and images if the corresponding loaders are included. webpack takes modules with dependencies and generates static assets representing those modules.",
  "to": "zh-CN",
  "trans": Object {
    "paragraphs": Array [
      "Webpack是一个开源的JavaScript模块绑定器。它主要是为JavaScript制作的，但如果包含相应的加载器，它可以转换前端资产，如HTML、CSS和图像。Webpack接受依赖的模块，并生成代表这些模块的静态资产。",
    ],
  },
}
`;
